validator/coordinator/src/main/java/tech/pegasys/teku/validator/coordinator/performance/DefaultPerformanceTracker.java

```diff
@@ -0,0 +1,269 @@
+import static tech.pegasys.teku.datastructures.util.BeaconStateUtil.compute_start_slot_at_epoch;
+import static tech.pegasys.teku.datastructures.util.BeaconStateUtil.get_block_root;
+import static tech.pegasys.teku.datastructures.util.BeaconStateUtil.get_block_root_at_slot;
+
+import com.google.common.annotations.VisibleForTesting;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.IntSummaryStatistics;
+import java.util.List;
+import java.util.Map;
+import java.util.NavigableMap;
+import java.util.Optional;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.stream.Collectors;
+import org.apache.tuweni.bytes.Bytes32;
+import tech.pegasys.teku.datastructures.blocks.BeaconBlock;
+import tech.pegasys.teku.datastructures.blocks.SignedBeaconBlock;
+import tech.pegasys.teku.datastructures.operations.Attestation;
+import tech.pegasys.teku.datastructures.state.BeaconState;
+import tech.pegasys.teku.infrastructure.logging.StatusLogger;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+import tech.pegasys.teku.ssz.SSZTypes.Bitlist;
+import tech.pegasys.teku.storage.client.CombinedChainDataClient;
+import tech.pegasys.teku.util.config.Constants;
+
+public class DefaultPerformanceTracker implements PerformanceTracker {
+
+  @VisibleForTesting
+  final NavigableMap<UInt64, Set<SignedBeaconBlock>> producedBlocksByEpoch = new TreeMap<>();
+
+  @VisibleForTesting
+  final NavigableMap<UInt64, Set<Attestation>> producedAttestationsByEpoch = new TreeMap<>();
+
+  @VisibleForTesting
+  static final UInt64 BLOCK_PERFORMANCE_EVALUATION_INTERVAL = UInt64.valueOf(2); // epochs
+
+  static final UInt64 ATTESTATION_INCLUSION_RANGE = UInt64.valueOf(2);
+
+  private final CombinedChainDataClient combinedChainDataClient;
+  private final StatusLogger statusLogger;
+  private final ValidatorPerformanceMetrics validatorPerformanceMetrics;
+
+  private Optional<UInt64> nodeStartEpoch = Optional.empty();
+  private AtomicLong latestAnalyzedEpoch = new AtomicLong(0);
+
+  public DefaultPerformanceTracker(
+      CombinedChainDataClient combinedChainDataClient,
+      StatusLogger statusLogger,
+      ValidatorPerformanceMetrics validatorPerformanceMetrics) {
+    this.combinedChainDataClient = combinedChainDataClient;
+    this.statusLogger = statusLogger;
+    this.validatorPerformanceMetrics = validatorPerformanceMetrics;
+  }
+
+  @Override
+  public void start(UInt64 nodeStartSlot) {
+    this.nodeStartEpoch = Optional.of(compute_epoch_at_slot(nodeStartSlot));
+  }
+
+  @Override
+  public void onSlot(UInt64 slot) {
+    if (nodeStartEpoch.isEmpty()) {
+      return;
+    }
+
+    if (slot.mod(Constants.SLOTS_PER_EPOCH).isGreaterThan(UInt64.ZERO)) {
+      return;
+    }
+
+    UInt64 currentEpoch = compute_epoch_at_slot(slot);
+    if (currentEpoch.longValue() <= latestAnalyzedEpoch.getAndSet(currentEpoch.longValue())) {
```

Comment:
I'd suggest using an `AtomicReference<UInt64>` so we don't have to mess with the long value conversion:
```suggestion
    if (currentEpoch.isLessThanOrEqualTo(latestAnalyzedEpoch.getAndSet(currentEpoch))) {
```

Reply:
Even better, you can use `getAndUpdate` to make sure `latestAnalyzedEpoch` only ever increases:
```suggestion
    if (currentEpoch.isLessThanOrEqualTo(latestAnalyzedEpoch.getAndUpdate(val -> val.isLessThan(currentEpoch) ? currentEpoch : val))) {
```

Reply:
That's a great suggestion. Done.